File: year2/week13/solution/week13.java
import java.util.*;
import java.util.function.*;
import java.util.concurrent.*;

public class week13 {
    public static void main(String args[]) {
        ActionBuffer ab = new ActionBuffer(5);
        for (int i = 0; i < 10; ++i) {
            Image im = new Image(i + "");
            im.load()
                .thenAccept((Image x) -> ab.defer(() -> System.out.println(x.path)));
        }
    }
}

interface Action extends Function<Void, Void> {
    default Void apply(Void v) {
        apply();
        return null;
    }
    void apply();
}

class ActionBuffer {
    ArrayDeque<Action> defered = new ArrayDeque<Action>();
    int waiting;

    public ActionBuffer(int n) {
        waiting = n;
    }

    public synchronized void defer(Action a) {
        defered.add(a);
        if (defered.size() == waiting)
            flush();
    }

    public synchronized void flush() {
        while (!defered.isEmpty())
            defered.remove().apply();
    }
}

class Image {
    public String path;

    public Image(String path) {
        this.path = path;
    }

    CompletableFuture<Image> load() {
        return CompletableFuture.supplyAsync(() -> {
            try {
                Thread.sleep(new Random().nextInt(4000));
            } catch (Exception e) {}
            return this;
        });
    }
}

File: year2/week13/solution/week13.js
function actionBuffer(n) {
    var self    = this;
    var waiting = n;
    const reset = n;
    var defered = [];

    this.defer = function(f) {
        --waiting;
        defered.push(f);
        if (waiting == 0) {
            self.flush();
            waiting = reset;
        }
    }

    this.flush = function() {
        while (defered.length != 0) {
            defered.shift()();
        }
    }
}

function bind(f, args) {
    return function() {
        f.apply(null, args);
    };
}

q = new actionBuffer(5);
for (var i = 0; i < 5; ++i) {
    console.log("queueing ", i);
    q.defer(bind(console.log, [i]));
}
